In [1]:
import numpy as np

## Ejercicio 1

Crea una array de numpy con los siguiente números primos: ```2, 3, 5, 7, 11, 13, 17, 19, 23, 29```

In [3]:
a1 = np.array([2, 3, 5, 7, 11, 13, 17, 19, 23, 29])

Muestra el último elemento. Y debes hacerlo sin saber su longitud.

In [5]:
print(a1[-1])

29


Muestra el 3º elemento

In [7]:
print(a1[2])

5


Muestra el último y 3º elemento.

In [9]:
print(a1[-1], a1[2])

29 5


Muestra del 2º al 5º elemento.

In [11]:
print(a1[1:5])

[ 3  5  7 11]


Muestra el 2° y el 5° elemento

In [13]:
print(a1[1], a1[4])

3 11


Muestra el último y el penúltimo elemento. Y debes hacerlo sin saber su longitud.

In [15]:
print(a1[-1], a1[-2])

29 23


# Ejercicio 2 
Crea una array de numpy con los siguiente números primos: ```2, 3, 5, 7, 11, 13, 17, 19, 23, 29``` pero que el array sea de tipo "float" de 32 bits.

In [17]:
a2_1 = np.array([2, 3, 5, 7, 11, 13, 17, 19, 23, 29], dtype=np.float32)
print(a2_1)

[ 2.  3.  5.  7. 11. 13. 17. 19. 23. 29.]


Usando la función arange de numpy crea un array con los números del 100 al 200 pero sin incluir el 200. Muestra el resultado.

In [19]:
a2_2 = np.arange(100, 200)
print(a2_2)

[100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117
 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135
 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153
 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171
 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189
 190 191 192 193 194 195 196 197 198 199]


Usando la función range estándar de Python crea un array con los números del 100 al 200 pero sin incluir el 200. Muestra el resultado. ¿Qué diferencia hay con arange?

In [21]:
a2_3 = list(range(100, 200))
print(a2_3)

# Range crea una lista nativa de Python

[100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199]


# Ejercicio 3
Crea la siguiente matriz: 
```python
[[1, 4, 6, 5]
 [4, 1, 7, 3]
 [2, 9, 1, 2]
 [6, 3, 1, 1]]
```

In [23]:
a3 = np.array([[1, 4, 6, 5],[4, 1, 7, 3],[2, 9, 1, 2],[6, 3, 1, 1]])

Muestra el elemento de la fila 2º y la columna 3º. Es el valor del 7.

In [25]:
print(a3[1, 2])

7


Muestra la 3º Fila

In [27]:
print(a3[1])

[4 1 7 3]


Muestra la 2º Columna

In [29]:
print(a3[:, 1])

[4 1 9 3]


Muestra la 2º y 3º Columna

In [31]:
print(a3[:, 1:3])

[[4 6]
 [1 7]
 [9 1]
 [3 1]]


Muestra la 2º y 3º Fila

In [33]:
print(a3[1:3])

[[4 1 7 3]
 [2 9 1 2]]


Muestra la última columna. Debe funcionar independientemente del número de columnas.

In [35]:
print(a3[:, -1])

[5 3 2 1]


Muestra la 2º y 4º Columna y la 1º y 3º fila

In [37]:
print("Columnas:")
print(a3[:,1:5:2])
print("Filas:")
print(a3[:4:2, :])

Columnas:
[[4 5]
 [1 3]
 [9 2]
 [3 1]]
Filas:
[[1 4 6 5]
 [2 9 1 2]]


Muestra de la 2º a la 3º Columna y de la 1º a la 3º fila

In [39]:
print("Columnas:")
print(a3[:,1:3])
print("Filas:")
print(a3[:3, :])

Columnas:
[[4 6]
 [1 7]
 [9 1]
 [3 1]]
Filas:
[[1 4 6 5]
 [4 1 7 3]
 [2 9 1 2]]


Muestra todas las columnas excepto la primera y la última. Debe funcionar independientemente del número de columnas.

In [41]:
print(a3[:, 1:-1])

[[4 6]
 [1 7]
 [9 1]
 [3 1]]


Muestra todas las filas excepto la primera y la última. Debe funcionar independientemente del número de filas.

In [43]:
print(a3[1:-1])

[[4 1 7 3]
 [2 9 1 2]]


Muestra todas las columnas excepto la primera y la última y todas las filas excepto la primera y la última. Debe funcionar independientemente del número de filas y columnas.

In [45]:
print("Columnas:")
print(a3[:, 1:-1])
print("Filas:")
print(a3[1:-1])

Columnas:
[[4 6]
 [1 7]
 [9 1]
 [3 1]]
Filas:
[[4 1 7 3]
 [2 9 1 2]]


Imprime la matriz y haz que las cabeceras de cada columna sean A, B , C y D

In [47]:
# Definir las cabeceras
headers = ['A', 'B', 'C', 'D']

# Imprimir las cabeceras
print('  ' + ' '.join(headers))

# Imprimir las filas de la matriz
print(a3)
#for row in a3:
    #print('  '.join(map(str, row)))

  A B C D
[[1 4 6 5]
 [4 1 7 3]
 [2 9 1 2]
 [6 3 1 1]]


# Ejercicio 4
El siguiente array contiene las temperaturas medias que ha hecho en Valencia en cada mes 
```python
[10.2, 10.7, 13.3, 15.8, 19.3, 23.6, 26, 25.9, 22.8, 19.1, 13.9, 10.8]
```

In [49]:
a4 = np.array([10.2, 10.7, 13.3, 15.8, 19.3, 23.6, 26, 25.9, 22.8, 19.1, 13.9, 10.8])

Muestra las temperaturas cuyo valor sea mayor que 20

In [51]:
print(a4[a4 > 20])

[23.6 26.  25.9 22.8]


Muestra las temperaturas cuyo valor sea menor que 11

In [53]:
print(a4[a4 < 11])

[10.2 10.7 10.8]


Muestra las temperaturas cuyo valor sea mayor que 20 o menor que 11

In [55]:
print(a4[(a4 > 20) | (a4 < 11)])

[10.2 10.7 23.6 26.  25.9 22.8 10.8]


# Ejercicio 5
Carga los datos de un dataset muy utilizado en la ciencia de datos con con el siguiente código:
```python
from sklearn.datasets import load_iris
datos=load_iris().data
resultado=load_iris().target
```

In [57]:
from sklearn.datasets import load_iris
datos=load_iris().data
resultado=load_iris().target

Crea un array llamado sepal_length con las 99 primeras filas y la 1º columna de la matriz datos

In [59]:
sepal_length = datos[:99, 0]
print(sepal_length)

[5.1 4.9 4.7 4.6 5.  5.4 4.6 5.  4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.  5.  5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.
 5.5 4.9 4.4 5.1 5.  4.5 4.4 5.  5.1 4.8 5.1 4.6 5.3 5.  7.  6.4 6.9 5.5
 6.5 5.7 6.3 4.9 6.6 5.2 5.  5.9 6.  6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
 6.3 6.1 6.4 6.6 6.8 6.7 6.  5.7 5.5 5.5 5.8 6.  5.4 6.  6.7 6.3 5.6 5.5
 5.5 6.1 5.8 5.  5.6 5.7 5.7 6.2 5.1]


Crea un array llamado petal_length con las 99 primeras filas y la 3º columna de la matriz datos

In [61]:
petal_length = datos[:99, 2]
print(petal_length)

[1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4 1.5 1.5 1.6 1.4 1.1 1.2 1.5 1.3 1.4
 1.7 1.5 1.7 1.5 1.  1.7 1.9 1.6 1.6 1.5 1.4 1.6 1.6 1.5 1.5 1.4 1.5 1.2
 1.3 1.4 1.3 1.5 1.3 1.3 1.3 1.6 1.9 1.4 1.6 1.4 1.5 1.4 4.7 4.5 4.9 4.
 4.6 4.5 4.7 3.3 4.6 3.9 3.5 4.2 4.  4.7 3.6 4.4 4.5 4.1 4.5 3.9 4.8 4.
 4.9 4.7 4.3 4.4 4.8 5.  4.5 3.5 3.8 3.7 3.9 5.1 4.5 4.5 4.7 4.4 4.1 4.
 4.4 4.6 4.  3.3 4.2 4.2 4.2 4.3 3. ]


Crea un array llamado x juntando las 2 columnas sepal_length y petal_length

In [63]:
x = np.concatenate((sepal_length, petal_length))
print(x)

[5.1 4.9 4.7 4.6 5.  5.4 4.6 5.  4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.  5.  5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.
 5.5 4.9 4.4 5.1 5.  4.5 4.4 5.  5.1 4.8 5.1 4.6 5.3 5.  7.  6.4 6.9 5.5
 6.5 5.7 6.3 4.9 6.6 5.2 5.  5.9 6.  6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
 6.3 6.1 6.4 6.6 6.8 6.7 6.  5.7 5.5 5.5 5.8 6.  5.4 6.  6.7 6.3 5.6 5.5
 5.5 6.1 5.8 5.  5.6 5.7 5.7 6.2 5.1 1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4
 1.5 1.5 1.6 1.4 1.1 1.2 1.5 1.3 1.4 1.7 1.5 1.7 1.5 1.  1.7 1.9 1.6 1.6
 1.5 1.4 1.6 1.6 1.5 1.5 1.4 1.5 1.2 1.3 1.4 1.3 1.5 1.3 1.3 1.3 1.6 1.9
 1.4 1.6 1.4 1.5 1.4 4.7 4.5 4.9 4.  4.6 4.5 4.7 3.3 4.6 3.9 3.5 4.2 4.
 4.7 3.6 4.4 4.5 4.1 4.5 3.9 4.8 4.  4.9 4.7 4.3 4.4 4.8 5.  4.5 3.5 3.8
 3.7 3.9 5.1 4.5 4.5 4.7 4.4 4.1 4.  4.4 4.6 4.  3.3 4.2 4.2 4.2 4.3 3. ]


Crea un array llamado y con las 99 primeras filas del vector resultado

In [65]:
y = resultado[:99]
print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


Crea un array llamado z con las 99 primeras filas , la 1º columna y la 3º columna de la matriz datos.

In [67]:
z = datos[:99,::2]
print(z)

[[5.1 1.4]
 [4.9 1.4]
 [4.7 1.3]
 [4.6 1.5]
 [5.  1.4]
 [5.4 1.7]
 [4.6 1.4]
 [5.  1.5]
 [4.4 1.4]
 [4.9 1.5]
 [5.4 1.5]
 [4.8 1.6]
 [4.8 1.4]
 [4.3 1.1]
 [5.8 1.2]
 [5.7 1.5]
 [5.4 1.3]
 [5.1 1.4]
 [5.7 1.7]
 [5.1 1.5]
 [5.4 1.7]
 [5.1 1.5]
 [4.6 1. ]
 [5.1 1.7]
 [4.8 1.9]
 [5.  1.6]
 [5.  1.6]
 [5.2 1.5]
 [5.2 1.4]
 [4.7 1.6]
 [4.8 1.6]
 [5.4 1.5]
 [5.2 1.5]
 [5.5 1.4]
 [4.9 1.5]
 [5.  1.2]
 [5.5 1.3]
 [4.9 1.4]
 [4.4 1.3]
 [5.1 1.5]
 [5.  1.3]
 [4.5 1.3]
 [4.4 1.3]
 [5.  1.6]
 [5.1 1.9]
 [4.8 1.4]
 [5.1 1.6]
 [4.6 1.4]
 [5.3 1.5]
 [5.  1.4]
 [7.  4.7]
 [6.4 4.5]
 [6.9 4.9]
 [5.5 4. ]
 [6.5 4.6]
 [5.7 4.5]
 [6.3 4.7]
 [4.9 3.3]
 [6.6 4.6]
 [5.2 3.9]
 [5.  3.5]
 [5.9 4.2]
 [6.  4. ]
 [6.1 4.7]
 [5.6 3.6]
 [6.7 4.4]
 [5.6 4.5]
 [5.8 4.1]
 [6.2 4.5]
 [5.6 3.9]
 [5.9 4.8]
 [6.1 4. ]
 [6.3 4.9]
 [6.1 4.7]
 [6.4 4.3]
 [6.6 4.4]
 [6.8 4.8]
 [6.7 5. ]
 [6.  4.5]
 [5.7 3.5]
 [5.5 3.8]
 [5.5 3.7]
 [5.8 3.9]
 [6.  5.1]
 [5.4 4.5]
 [6.  4.5]
 [6.7 4.7]
 [6.3 4.4]
 [5.6 4.1]
 [5.5 4. ]
 [5.5 4.4]

# Ejercicio 6
Selecciona las celdas en rojo oscuro de la siguiente matriz:

![Matriz 1](./matriz1.png)

```
[[ 1  2  3  4  5  6  7  8]
 [ 9 10 11 12 13 14 15 16]
 [17 18 19 20 21 22 23 24]
 [25 26 27 28 29 30 31 32]
 [33 34 35 36 37 38 39 40]
 [41 42 43 44 45 46 47 48]
 [49 50 51 52 53 54 55 56]]
```

In [69]:
a6 = np.arange(1, 57)
a6 = a6.reshape(7, 8)

print(a6[2:5, 2:6])

[[19 20 21 22]
 [27 28 29 30]
 [35 36 37 38]]


# Ejercicio 7
Selecciona las celdas en rojo oscuro de la siguiente matriz:

![Matriz 2](./matriz2.png)

```
[[[  1   2   3   4   5   6   7   8]
  [  9  10  11  12  13  14  15  16]
  [ 17  18  19  20  21  22  23  24]
  [ 25  26  27  28  29  30  31  32]
  [ 33  34  35  36  37  38  39  40]
  [ 41  42  43  44  45  46  47  48]
  [ 49  50  51  52  53  54  55  56]]

 [[ 57  58  59  60  61  62  63  64]
  [ 65  66  67  68  69  70  71  72]
  [ 73  74  75  76  77  78  79  80]
  [ 81  82  83  84  85  86  87  88]
  [ 89  90  91  92  93  94  95  96]
  [ 97  98  99 100 101 102 103 104]
  [105 106 107 108 109 110 111 112]]

 [[113 114 115 116 117 118 119 120]
  [121 122 123 124 125 126 127 128]
  [129 130 131 132 133 134 135 136]
  [137 138 139 140 141 142 143 144]
  [145 146 147 148 149 150 151 152]
  [153 154 155 156 157 158 159 160]
  [161 162 163 164 165 166 167 168]]]
```

In [71]:
a7 = np.arange(1, 169)
a7 = a7.reshape(3, 7, 8)

print(a7[0, 2:5, 2:6])

[[19 20 21 22]
 [27 28 29 30]
 [35 36 37 38]]


Ahora selecciona las celdas en rojo oscuro pero también las verdes y azules que hay por detrás de las rojo oscuro.

In [73]:
print(a7[:, 2:5, 2:6])

[[[ 19  20  21  22]
  [ 27  28  29  30]
  [ 35  36  37  38]]

 [[ 75  76  77  78]
  [ 83  84  85  86]
  [ 91  92  93  94]]

 [[131 132 133 134]
  [139 140 141 142]
  [147 148 149 150]]]


# Ejercicio 8
Crea la siguiente matriz:
```python
[[1, 4, 6, 5]
 [4, 1, 7, 3]
 [2, 9, 1, 2]
 [6, 3, 1, 1]]
```

In [75]:
a8 = np.array([[1, 4, 6, 5],[4, 1, 7, 3],[2, 9, 1, 2],[6, 3, 1, 1]])

A todos los elementos de la matriz sumale un 10. Usando las funciones de numpy

In [77]:
print(np.add(a8, 10))

[[11 14 16 15]
 [14 11 17 13]
 [12 19 11 12]
 [16 13 11 11]]


A todos los elementos de la matriz sumale un 10 y divídelos entre 2. Usando las funciones de numpy

In [139]:
print(np.divide(np.add(a8, 10), 2))

[[5.5 7.  8.  7.5]
 [7.  5.5 8.5 6.5]
 [6.  9.5 5.5 6. ]
 [8.  6.5 5.5 5.5]]


A todos los elementos de la matriz sumale un 10. Usando operadores

In [142]:
print(a8 + 10)

[[11 14 16 15]
 [14 11 17 13]
 [12 19 11 12]
 [16 13 11 11]]


A todos los elementos de la matriz sumale un 10 y divídelos entre 2. Usando operadores

In [145]:
print((a8 +10)/2)

[[5.5 7.  8.  7.5]
 [7.  5.5 8.5 6.5]
 [6.  9.5 5.5 6. ]
 [8.  6.5 5.5 5.5]]


# Ejercicio 9
Crea la siguiente matriz
```python
[[1, 2, 3]
 [4, 5, 6]
 [7, 8, 9]]
```

Multiplica cada elemento de la matriz por sí mismo
Multiplica la matriz por sí misma

# Ejercicio 10
Crea una función llamada f que acepte como parámetro el número y que retorne el valor multiplicado por 2 y además que se le sume 1.
Crea ahora el vector de numpy [1,5,4,7,3,9,8,6] y aplícale la función f

# Ejercicio 11
Crea una función llamada f que acepte como parámetro el número y que retorne lo siguiente:
Si es valor está en el rango (-inf,3] que retorne 0
Si es valor está en el rango [3,5] que retorne 4
Si es valor está en el rango [5,6] que retorne 5
Si es valor está en el rango [6,7] que retorne 6
Si es valor está en el rango [7,9] que retorne 8
Si es valor está en el rango [9,+inf) que retorne 10
Crea ahora el vector de numpy [1,5,4,7,3,9,8,6] y aplícale la función f

# Ejercicio 12
Muestra con numpy el número de dimensiones de los siguientes arrays:


a=[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37]




b=[[2, 3], [5, 7], [11, 13], [17, 19], [23, 29], [31, 37]]




c=[[[2, 3], [5, 7]], [[11, 13], [17, 19]], [[23, 29], [31, 37]]]

# Ejercicio 13: Forma
Crea una función llamada mostrar_tamanyo que le pasemos un array de numpy y nos imprima el tamaño de cada una de las dimensiones. 
Por ejemplo con el array [ [ [2, 3], [5, 7]], [ [11, 13], [17, 19]], [ [23, 29], [31, 37] ] ] deberá mostrar
El nº de elementos de la dimension 0 es 3
El nº de elementos de la dimension 1 es 2
El nº de elementos de la dimension 2 es 2

Prueba también con los siguientes arrays


a=[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37]




b=[[2, 3], [5, 7], [11, 13], [17, 19], [23, 29], [31, 37]]




c=[[[2, 3], [5, 7]], [[11, 13], [17, 19]], [[23, 29], [31, 37]]]

# Ejercicio 14
Muestra con numpy el tipo de datos de los siguientes arrays:


a=[2, 3, 5]




b=[2.4, 3.2, 5.6]




c=[2, 3, 5.6]

# Ejercicio 15
Transforma el siguiente array en 8 bits sin signo y muestra cuánto ocupa en memoria


a=np.array([1, -2, 3])


Transforma el siguiente array en 8 bits sin signo y muestra cuánto ocupa en memoria


b=np.array([1,254,255])


Transforma el siguiente array en 8 bits con signo y muestra cuánto ocupa en memoria


b=np.array([1,127,128])


Transforma el siguiente array en 8 bits con signo y muestra cuánto ocupa en memoria


b=np.array([1,-128,-129])

Transforma el siguiente array en float32 y muestra cuanto ocupa en memoria


b=np.array([1,2,3])


Transforma el siguiente array en int32 y muestra cuánto ocupa en memoria


b=np.array([1.9,2.01,3.51])

# Ejercicio 16
En ejercicios anteriores obtuviste para el problema de las flores, una matriz x y un vector y.
Une los datos en una nueva matriz de forma que cada fila de la nueva matriz contenga los datos de cada fila de la x y el dato correspondiente de la y.

# Ejercicio 17
Dado el siguiente código python:

```python
import matplotlib.pyplot as plt
import numpy as np
figure=plt.figure(figsize=(8,8))
axes = figure.add_subplot()
x=[-3,-2,-1,0,1,2,3]
y =  3*(1 - x)**2 * np.exp(-x**2 )  - 10*(x/5 - x**3 )*np.exp(-x**2 ) - 1./3*np.exp(-(x + 1)**2 )
axes.plot(x,y)
```


Modifica el código que genera la variable x para que sea un array de numpy y de tipo float. Muestra la imagen
Modifica el código que genera la variable x para que sean 10 valores entre el [-3,3]. Muestra la imagen
Modifica el código que genera la variable x para que sean 20 valores entre el [-3,3]. Muestra la imagen
Modifica el código que genera la variable x para que sean 40 valores entre el [-3,3]. Muestra la imagen
Modifica el código que genera la variable x para que sean 60 valores entre el [-3,3]. Muestra la imagen
Modifica el código que genera la variable x para que sean 100 valores entre el [-3,3]. Muestra la imagen

# Ejercicio 18
Carga con numpy el fichero wine.csv que contiene una matriz de números separados por coma.

La matriz contiene una serie de columnas con características del vino y la última columna indica el tipo del vino.
Crea una función en python llamada imprimir_datos y tendrá como parámetro una matriz de numpy de forma que:
Imprima cuantas características tenemos de cada tipo de vino.
Imprima cuantas muestra de vino tenemos
Imprima cuantos tipos distintos de vino tenemos y cuales son
Imprima los valores máximos y mínimos de cada característica
Retorna una nueva matriz llamada x solo con las características y un nuevo array llamado y solo con los tipos de vino
Llama a la función imprimir_datos con los datos de wine.csv

# Ejercicio 19

Tienes los datos de las ventas diarias de una tienda durante un año. Los datos incluyen ventas anormalmente altas debido a días festivos o promociones. El objetivo es limpiar los datos eliminando estos valores atípicos y analizar la tendencia general de ventas.


```python
np.random.seed(0)
# Ventas diarias simuladas con media 500 y desviación estándar 50
sales_data = np.random.normal(500, 50, 365)  
# Ventas atípicas por promociones
sales_data[[50, 120, 180, 250, 300]] = [1500, 1400, 1600, 1700, 1800]  
```


Detecta y elimina los valores atípicos (aquellos que están fuera de 3 desviaciones estándar del promedio).
Calcula el promedio, la mediana y la desviación estándar de las ventas después de limpiar los datos.
Genera una vista de las ventas de cada trimestre (3 primeros meses, 3 siguientes, etc.).
Encuentra los 5 días con las ventas más bajas después de la limpieza de los datos.

# Ejercicio 20

Tienes un conjunto de datos que contiene información demográfica de varias personas, incluyendo edad, ingresos anuales, y nivel de educación. El objetivo es realizar un análisis y encontrar patrones significativos entre estas variables.

```python
np.random.seed(3)
ages = np.random.randint(18, 70, size=200)  # Edades de 200 personas
income = np.random.normal(50000, 15000, size=200)  # Ingresos anuales en un rango normal alrededor de 50,000
education_years = np.random.randint(0, 20, size=200)  # Años de educación formal
```

Calcula la correlación entre los años de educación y los ingresos.
Divide a las personas en grupos de edades (18-30, 31-45, 46-60, 61+) y calcula el ingreso promedio por grupo.
Encuentra el percentil 90 de ingresos y determina cuántas personas ganan por encima de ese valor.
Agrupa los datos en función del número de años de educación y calcula el promedio de ingresos para cada grupo.

Para calcular la correlación utiliza la función corrcoef.

Para encontrar el percentil 90 utiliza la función percentile.